### ISSUE TITLE
Inconsistent output of Find All References 

### ISSUE THREAD

**Issue by vladzcloudius** [microsoft/vscode-cpptools#5376]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Ubuntu 19.10
- VS Code Version: 
```
Version: 1.44.2
Commit: ff915844119ce9485abfe8aa9076ec76b5300ddd
Date: 2020-04-16T17:50:03.709Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Linux x64 5.6.6-050606-generic snap
```
- C/C++ Extension Version: `Version 0.27.0: March 30, 2020`
- Other extensions you installed (and if the issue persists after disabling them): None
- Does this issue involve using SSH remote to run the extension on a remote machine?: No
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

**Steps to reproduce**

1. `git clone https://github.com/scylladb/scylla.git` 
2. cd scylla
3. git checkout 1f902302ad8d26f9db8c0622b3bcc844f07d4517
4. git submodule update --init --recursive
5. code .
6. Open `utils/loading_cache.hh`
7. Go to line 254: https://github.com/scylladb/scylla/blob/master/utils/loading_cache.hh#L254
8. Find all references of class `loading_cache`.

It finds all but one reference (in `authorized_prepared_statements_cache.hh`):

![image](https://user-images.githubusercontent.com/6103981/80542153-86467000-897a-11ea-8c22-2e3b6e730ed2.png)

9. Go to a loading_cache class reference in the `authorized_prepared_statements_cache.hh` and look for all references again.
This time it is going to find them all:

![image](https://user-images.githubusercontent.com/6103981/80542297-bdb51c80-897a-11ea-84dd-ab0150814371.png)

10. Go back to loading_cache.hh and try to search for all references again. It is going to find no references at all:

![image](https://user-images.githubusercontent.com/6103981/80542817-b93d3380-897b-11ea-87a3-61ad56587ad3.png)

At the same time Go To Definition always works as expected.

**Expected behavior**

It should find all references all times as in step (9) above.

1. Logs from the command `C/C++: Log Diagnostics`:

```
-------- Diagnostics - 4/28/2020, 6:13:46 PM
Version: 0.27.0
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/home/vladz/.local/bin/gcc",
    "intelliSenseMode": "clang-x64",
    "compilerArgs": [],
    "cStandard": "c11",
    "cppStandard": "c++20",
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ]
    }
}
Translation Unit Mappings:
[ /home/vladz/work/tmp/scylla/cql3/authorized_prepared_statements_cache.hh ]:
    /home/vladz/work/tmp/scylla/cql3/authorized_prepared_statements_cache.hh
    /home/vladz/work/tmp/scylla/utils/loading_cache.hh
Translation Unit Configurations:
[ /home/vladz/work/tmp/scylla/cql3/authorized_prepared_statements_cache.hh ]:
    Process ID: 31166
    Memory Usage: 1161 MB
    Compiler Path: /home/vladz/.local/bin/gcc
    Includes:
        /usr/include/c++/9
        /usr/include/x86_64-linux-gnu/c++/9
        /usr/include/c++/9/backward
        /usr/lib/gcc/x86_64-linux-gnu/9/include
        /usr/local/include
        /usr/lib/gcc/x86_64-linux-gnu/9/include-fixed
        /usr/include/x86_64-linux-gnu
        /usr/include
        /home/vladz/work/tmp/scylla
        /home/vladz/work/tmp/scylla/seastar/include
        /home/vladz/work/tmp/scylla/zstd/tests/regression
        /home/vladz/work/tmp/scylla/zstd/contrib/linux-kernel/include/linux
    Standard Version: c++20
    IntelliSense Mode: gcc-x64
    Other Flags:
        --g++
        --gnu_version=90201
        --header_only_fallback
Total Memory Usage: 1161 MB

```

2. Logs from [the language server](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server):

Below is the log with a `debug` logging level when I tried to follow steps 7-10 above. However the tool was unable to find any reference in step 9.
(It looks like we were hitting some timeouts here. We may be hitting same timeouts in step 10 above too).

[logs.txt](https://github.com/microsoft/vscode-cpptools/files/4548909/logs.txt)

**Screenshots**
See above.

**Comment by vladzcloudius** [comment]
I tried increasing the Intelli Sense cache size in the workspace or global config - it didn't help.

**Comment by sean-mcmanus** [comment]
If you do a Find All References instead of a Peek you should be able to see the missing references in the Cannot Confirm References section (not sure of the reason for that yet)...peek also should show the Other References but there might be some bug preventing that some times.

One reason for the randomness is that for the header files, a different TU could be chosen on different runs, and some TUs may crash or hit some compiler error that prevents the reference from being confirmed. It would be easier to diagnose if we exposed what TU a header was being associated with (and more info when a Find All References fails to confirm references).

I may be able to find more info later...

**Comment by vladzcloudius** [comment]
> If you do a Find All References instead of a Peek you should be able to see the missing references in the Cannot Confirm References section (not sure of the reason for that yet)...peek also should show the Other References but there might be some bug preventing that some times.

That seems to work. Thanks.

> 
> One reason for the randomness is that for the header files, a different TU could be chosen on different runs, and some TUs may crash or hit some compiler error that prevents the reference from being confirmed. It would be easier to diagnose if we exposed what TU a header was being associated with (and more info when a Find All References fails to confirm references).

Please, let me know if you need more info from my end.


> 
> I may be able to find more info later...

**Comment by sean-mcmanus** [comment]
How do I get the missing "json/json.h" files?

**Comment by sean-mcmanus** [comment]
Oh, I figured it out -- from jsoncpp.

**Comment by sean-mcmanus** [comment]
I did some more investigation -- abstract_marker.cc is getting chosen as the TU for authorized_prepared_statements_cache.hh during Find All References when starting from loading_cache.hh. That file's TU gets created fine; however, if you open authorized_prepared_statements_cache.hh afterwards (which causes it to re-use the existing TU), IntelliSense for authorized_prepared_statements_cache.hh gets stuck in a "pending" state forever, the flame icon never disappears, which could be the root cause of the Find All References confirmation failure.

**Comment by sean-mcmanus** [comment]
For 0.28.0-insiders2, the case where the authorized_prepared_statements_cache.hh is opened after abstract_marker.cc is fixed, but your original repro via Find All References required a few more changes and didn't make it in in time, but it should be fixed in the following release (0.28.0-insiders3 or 0.28.0). As in, all the "Cannot Confirm References" are gone and the Find All References results are all the same in all 3 source locations.

**Comment by vladzcloudius** [comment]
> For 0.28.0-insiders2, the case where the authorized_prepared_statements_cache.hh is opened after abstract_marker.cc is fixed, but your original repro via Find All References required a few more changes and didn't make it in in time, but it should be fixed in the following release (0.28.0-insiders3 or 0.28.0). As in, all the "Cannot Confirm References" are gone and the Find All References results are all the same in all 3 source locations.

Thanks a lot, @sean-mcmanus!
Will be waiting for 0.28.0.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/0.28.0 .

I noticed our Find All References performance is very (unusually) slow with that repro, so it's something we may want to investigate more. I filed a new issue at https://github.com/microsoft/vscode-cpptools/issues/5485 .
